tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,681 @@
+    }
+
+    fn poll_acquire(
+        &self,
+        cx: &mut Context<'_>,
+        needed: u16,
+        node: Pin<&mut Waiter>,
+        queued: bool,
+    ) -> Poll<Result<(), AcquireError>> {
+        let mut acquired = 0;
+
+        // If we are already in the wait queue, we need to lock the queue so we
+        // can access the wait queue entry's current state.
+        let (needed, mut lock) = if queued {
+            let lock = self.waiters.lock().unwrap();
+            // Safety: since we have acquired the lock, it is safe to look at
+            // the waiter's state.
+            let needed = node.state.with(|curr| unsafe { *curr });
+            (needed, Some(lock))
+        } else {
+            (needed as usize, None)
+        };
+
+        // First, try to take the requested number of permits from the
+        // semaphore.
+        let mut curr = self.permits.load(Acquire);
+        let mut waiters = loop {
+            // Has the semaphore closed?
+            if curr & CLOSED > 0 {
+                return Ready(Err(AcquireError::closed()));
+            }
+
+            let mut remaining = 0;
+            let (next, acq) = if curr + acquired >= needed {
+                let next = curr - (needed - acquired);
+                (next, needed)
+            } else {
+                remaining = (needed - acquired) - curr;
+                (0, curr)
+            };
+
+            if remaining > 0 && lock.is_none() {
+                // No permits were immediately available, so this permit will
+                // (probably) need to wait. We'll need to acquire a lock on the
+                // wait queue.
+                if let Ok(l) = self.waiters.try_lock() {
+                    // If we were able to acquire the lock *now* without
+                    // blocking, we don't need to reload the semaphore's permit state.
+                    lock = Some(l)
+                } else {
+                    lock = Some(self.waiters.lock().unwrap());
+                    // While we were waiting to lock the wait list, additional
+                    // permits may have been released. Therefore, we will acquire a
+                    // new snapshot of the current state of the semaphore before
+                    // actually enqueueing the waiter..
+                    curr = self.permits.load(Acquire);
+                    continue;
+                }
+            }
+
+            match self
+                .permits
+                .compare_exchange_weak(curr, next, AcqRel, Acquire)
+            {
+                Ok(_) => {
+                    acquired += acq;
+                    if remaining == 0 && !queued {
+                        return Ready(Ok(()));
+                    }
+                    break lock.unwrap();
+                }
+                Err(actual) => curr = actual,
+            }
+            drop(lock.take());
+        };
+
+        if node.assign_permits(&mut acquired) {
+            if acquired > 0 {
```

Comment:
When does this happen? I know it did happen with the old `semaphore_ll`, but that was terrible!

Reply:
This didn't happen when the waiter state was atomic, since we could load it immediately on `poll_acquire`; when moving the waiter state into the lock, we started leaking permits. I _think_ this could be avoided if we acquired the lock earlier in `poll_acquire` when we know we are already queued, though.

A side note: [this `loom` test](https://github.com/tokio-rs/tokio/blob/c9699dc827212c4f2aa5a8c05fd954abbb0c2bd2/tokio/src/sync/tests/loom_semaphore_batch.rs#L175-L193) has caught about 2/3rds of the bugs in the implementation :)